presto-main/src/main/java/io/prestosql/sql/planner/DomainCoercer.java

```diff
@@ -0,0 +1,252 @@
+    {
+        private final ConnectorSession connectorSession;
+        private final InterpretedFunctionInvoker functionInvoker;
+        private final ResolvedFunction saturatedFloorCastOperator;
+        private final ResolvedFunction castToOriginalTypeOperator;
+        private final MethodHandle comparisonOperator;
+        private final Domain domain;
+        private final Type coercedValueType;
+
+        private ImplicitCoercer(Metadata metadata, TypeOperators typeOperators, Session session, Domain domain, Type coercedValueType)
+        {
+            this.connectorSession = requireNonNull(session, "session is null").toConnectorSession();
+            this.functionInvoker = new InterpretedFunctionInvoker(metadata);
+            this.domain = requireNonNull(domain, "domain is null");
+            this.coercedValueType = requireNonNull(coercedValueType, "coercedValueType is null");
+            Type originalValueType = domain.getType();
+            try {
+                this.saturatedFloorCastOperator = metadata.getCoercion(SATURATED_FLOOR_CAST, originalValueType, coercedValueType);
+            }
+            catch (OperatorNotFoundException e) {
+                throw new IllegalStateException(
+                        format("Saturated floor cast operator not found for coercion from %s to %s", originalValueType, coercedValueType));
+            }
+            this.castToOriginalTypeOperator = metadata.getCoercion(coercedValueType, originalValueType);
+            this.comparisonOperator = typeOperators.getComparisonOperator(originalValueType, InvocationConvention.simpleConvention(FAIL_ON_NULL, NEVER_NULL, NEVER_NULL));
+        }
+
+        public Domain applySaturatedCasts()
+        {
+            if (domain.isNone()) {
+                return Domain.none(coercedValueType);
+            }
+            ValueSet saturatedValueSet = domain.getValues().getValuesProcessor().transform(
+                    this::applySaturatedCasts,
+                    this::applySaturatedCasts,
+                    allOrNone -> new AllOrNoneValueSet(coercedValueType, allOrNone.isAll()));
+
+            return Domain.create(saturatedValueSet, domain.isNullAllowed());
+        }
+
+        private ValueSet applySaturatedCasts(Ranges ranges)
+        {
+            ImmutableList.Builder<Range> builder = ImmutableList.builder();
+            for (Range range : ranges.getOrderedRanges()) {
+                Optional<Range> coercedRange = applySaturatedCasts(range);
+                if (coercedRange.isEmpty()) {
+                    continue;
+                }
+                if (coercedRange.get().isAll()) {
+                    return ValueSet.all(coercedValueType);
+                }
+                builder.add(coercedRange.get());
+            }
+            return SortedRangeSet.copyOf(coercedValueType, builder.build());
+        }
+
+        private ValueSet applySaturatedCasts(DiscreteValues discreteValues)
```

Comment:
is it tested (or possible to test)?

Reply:
Doesn't seem possible to test in a meaningful way. Should we just drop this part and return `all` in this case to simplify things ?

Reply:
Let's drop it then.

Reply:
removed this now